File: src/devtwin/system/os.py
"""Operating system / platform inspection."""

from __future__ import annotations

import platform
from dataclasses import dataclass

from devtwin.core.models import DictMixin


@dataclass
class OSInfo(DictMixin):
    system: str  # "Darwin", "Linux", "Windows"
    release: str
    version: str
    machine: str
    python_version: str
    is_macos: bool
    is_linux: bool
    is_windows: bool


def get_os_info() -> OSInfo:
    system = platform.system()
    return OSInfo(
        system=system,
        release=platform.release(),
        version=platform.version(),
        machine=platform.machine(),
        python_version=platform.python_version(),
        is_macos=system == "Darwin",
        is_linux=system == "Linux",
        is_windows=system == "Windows",
    )

File: mcphub/devtwin/src/system/os.ts
/**
 * Operating system / platform inspection.
 */

import { arch, machine, release, version } from 'node:os';

export interface OsInfo {
  /** "Darwin", "Linux", "Windows" */
  system: string;
  release: string;
  version: string;
  machine: string;
  /** runtime version of the host process (the Node analogue of python_version) */
  node_version: string;
  is_macos: boolean;
  is_linux: boolean;
  is_windows: boolean;
}

/** Normalize `process.platform` onto the names Python's `platform.system()` reports. */
export function platformSystem(): string {
  switch (process.platform) {
    case 'darwin':
      return 'Darwin';
    case 'linux':
      return 'Linux';
    case 'win32':
      return 'Windows';
    default:
      return process.platform.charAt(0).toUpperCase() + process.platform.slice(1);
  }
}

export function getOsInfo(): OsInfo {
  const system = platformSystem();
  return {
    system,
    release: safe(() => release(), ''),
    version: safe(() => version(), ''),
    // os.machine() is the closest match to platform.machine(); arch() is the fallback.
    machine: safe(() => machine(), arch()),
    node_version: process.versions.node,
    is_macos: system === 'Darwin',
    is_linux: system === 'Linux',
    is_windows: system === 'Windows',
  };
}

/** Never crash on a platform where one of the os.* getters is unavailable. */
function safe(fn: () => string, fallback: string): string {
  try {
    return fn();
  } catch {
    return fallback;
  }
}
